src/pybind/mgr/cephadm/module.py

```diff
@@ -324,10 +333,194 @@ def __init__(self, *args, **kwargs):
+        self.event.set()
+
+    def _do_upgrade(self, daemons):
+        # type: (List[orchestrator.ServiceDescription]) -> Optional[AsyncCompletion]
+        if not self.upgrade_state:
+            self.log.debug('_do_upgrade no state, exiting')
+            return None
+
+        target_name = self.upgrade_state.get('target_name')
+        target_id = self.upgrade_state.get('target_id')
+        target_version = self.upgrade_state.get('target_version')
+        self.log.info('Upgrade: Target is %s id %s' % (target_name, target_id))
+
+        # get all distinct container_image settings
+        image_settings = {}
+        ret, out, err = self.mon_command({
+            'prefix': 'config dump',
+            'format': 'json',
+        })
+        config = json.loads(out)
+        for opt in config:
+            if opt['name'] == 'container_image':
+                image_settings[opt['section']] = opt['value']
+
+        for daemon_type in ['mgr', 'mon', 'osd', 'rgw', 'mds']:
+            self.log.info('Upgrade: Checking %s daemons...' % daemon_type)
+            need_upgrade_self = False
+            for d in daemons:
+                if d.service_type != daemon_type:
+                    continue
+                if daemon_type == 'mgr' and \
+                   d.service_instance == self.get_mgr_id():
+                    self.log.info('Upgrade: Need to upgrade myself (mgr.%s)' %
+                                  self.get_mgr_id())
+                    need_upgrade_self = True
+                    continue
+                if d.container_image_id != target_id:
+                    while True:
+                        if daemon_type not in ['mon', 'osd', 'mds']:
+                            break
+                        ret, out, err = self.mon_command({
+                            'prefix': '%s ok-to-stop',
+                            'ids': [d.service_instance],
+                        })
+                        if err:
+                            self.log.info('Upgrade: not safe to stop %s.%s' %
+                                          (daemon_type, d.service_instance))
+                            time.sleep(15)
+                            if not self.upgrade_state or self.upgrade_state.get('paused'):
+                                return
+                        else:
+                            self.log.info('Upgrade: safe to stop %s.%s' %
+                                          (daemon_type, d.service_instance))
+                            break
+
+                    self.log.info('Upgrade: Redeploying %s.%s' %
+                                  (d.service_type, d.service_instance))
+                    ret, out, err = self.mon_command({
+                        'prefix': 'config set',
+                        'name': 'container_image',
+                        'value': target_name,
+                        'who': daemon_type + '.' + d.service_instance,
+                    })
```

Comment:
Hm. My understanding of `container_image` was that it's meant to be used for specifying special (quick fix etc.) versions for some individual containers. As we're updating to the new default image, maybe we can simply rely on the default image? 

I.e. we simply don't write settings for individual daemons and instead set the new image as who=global upfront?